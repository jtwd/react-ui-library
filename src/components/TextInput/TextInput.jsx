import React from 'react'
import { string, oneOf, bool, func, node } from 'prop-types'
import Label from '../Label'
import Icon from '../Icon'
import { Field, Input, Error } from './TextInput.styles'

function TextInput({ htmlId, name, label, type, required, onChange, placeholder, value, error, children, ...props }) {
  return (
    <Field>
      <Label htmlFor={htmlId} label={label} required={required} />
      <Input
        id={htmlId}
        type={type}
        name={name}
        placeholder={placeholder}
        value={value}
        onChange={onChange}
        error={error}
        {...props}
      />
      {children}
      {error &&
        <Error>
          <Icon icon="exclamation" small />
          {error}
        </Error>}
    </Field>
  )
}

TextInput.propTypes = {
  /** ID for input to link with label */
  htmlId: string.isRequired,
  /** Name of input */
  name: string.isRequired,
  /** To go in label */
  label: string.isRequired,
  /** Input type (text, email, number, password, tel) */
  type: oneOf(['text', 'email', 'number', 'password', 'tel']),
  /** Required field */
  required: bool,
  /** Function to run when input is being typed in */
  onChange: func.isRequired,
  /** Placeholder text */
  placeholder: string,
  /** Value of input */
  // eslint-disable-next-line react/require-default-props
  value: string,
  /** Error message */
  error: string,
  children: node
}

TextInput.defaultProps = {
  type: 'text',
  required: false,
  placeholder: '',
  error: null,
  children: null
}

export default TextInput

