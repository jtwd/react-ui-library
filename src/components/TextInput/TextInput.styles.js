import styled, { css } from 'styled-components'

import { red, lightRed, yellow, white, transBlackDark, transBlackLight } from '../_theme/colors'
import { transTextHover, radiusRound } from '../_theme/units'
import { xxs, xs, md } from '../_theme/spacers'

export const Error = styled.div`
  color: ${lightRed};
  padding-top: ${xxs};
  
  > i {
    position: relative;
    top: 2px;
    left: -1px;
    margin-right: 6px;
  }
`

export const Field = styled.div`
  position: relative;
  margin-bottom: ${md};
`

export const Input = styled.input`
  border: 2px solid ${white};
  background: ${transBlackLight};
  color: ${white};
  padding: ${xs};
  transition: ${transTextHover};
  border-radius: ${radiusRound};
  display: block;
  width: 100%;
  
  &:focus {
    outline-color: ${yellow}
  }

  ${props =>
    props.error &&
    css`
    border-color: ${red};
  `} 
  
  &:focus {
    border-color: ${props => (props.error ? red : yellow)};
    background: ${transBlackDark};
    outline: 0;
  }
`
