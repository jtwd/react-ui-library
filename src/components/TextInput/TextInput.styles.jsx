import React from 'react'
import { string } from 'prop-types'
import styled from 'styled-components'

import { getInputStyles, fieldSizes } from "../_theme/forms"
import { alert } from '../_theme/palette'
import spacers from '../_theme/spacers'
import Message from '../Message'
import Icon from '../Icon'

const Error = styled(Message)`
  color: ${alert('light')};
  padding: 0;
  margin: 0;
  border: 0;
  background: transparent;
  box-shadow: none;
  &:hover {
    box-shadow: none;
  }
  
  ${props => props.active && `
    max-height: 100px;
  `}
  
  > div {
    padding: ${spacers.xxs} 0 ${spacers.xxs} ${spacers.lg};
    
    > i {
      margin-right: ${spacers.xxs};
      position: absolute;
      top: ${spacers.xs};
      left: ${spacers.xxs};
    }
  }
`

export function ErrorMsg({message}) {
  return (
    <Error active={(message)}>
      <Icon icon="exclamation" small />
      {message}
    </Error>
  )
}

ErrorMsg.propTypes = {
  message: string
}

ErrorMsg.defaultProps = {
  message: null
}

export const Field = styled.div`
  position: relative;
  margin-bottom: ${spacers.md};
  
  ${props => props.size !== 'default' && `
    max-width: ${fieldSizes[props.size]};
  `}
`

export const Input = styled.input`
  ${props => getInputStyles(props.error)}
`
