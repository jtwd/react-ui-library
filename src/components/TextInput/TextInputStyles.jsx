import React from 'react'
import { string } from 'prop-types'
import styled, { css } from 'styled-components'

import { red, lightRed, darkRed, yellow, white, transBlackDark, transBlackLight, transWhite } from '../_theme/colors'
import { transitions, radiusRound, formSizes } from '../_theme/units'
import { xxs, xs, md } from '../_theme/spacers'
import Message from '../Message'
import Icon from '../Icon'

const Error = styled(Message)`
  color: ${lightRed};
  padding: 0;
  margin: 0;
  border: 0;
  background: transparent;
  box-shadow: none;
  &:hover {
    box-shadow: none;
  }
  
  ${props => props.active && `
    max-height: 100px;
  `}
  
  > div {
    padding: ${xxs} 0 0;
    
    > i {
      margin-right: ${xxs};
      position: relative;
      top: 2px;
    }
  }
`

export function ErrorMsg({message}) {
  return (
    <Error active={(message)}>
      <Icon icon="exclamation" small />
      {message}
    </Error>
  )
}

ErrorMsg.propTypes = {
  message: string
}

ErrorMsg.defaultProps = {
  message: null
}

export const Field = styled.div`
  position: relative;
  margin-bottom: ${md};
  
  ${props => props.size !== 'default' && `
    max-width: ${formSizes[props.size]};
  `}
`

export const Input = styled.input`
  border: 2px solid ${transWhite};
  background: ${transBlackLight};
  color: ${white};
  padding: ${xs};
  transition: ${transitions.hover};
  border-radius: ${radiusRound};
  display: block;
  width: 100%;

  ${props =>
  props.error &&
  css`
    border-color: ${lightRed};
  `} 
  
  &:hover {
    border-color: ${props => (props.error ? red : white)};
  }
  
  &:focus {
    border-color: ${props => (props.error ? darkRed : yellow)};
    background: ${transBlackDark};
    outline: 0;
  }
`
