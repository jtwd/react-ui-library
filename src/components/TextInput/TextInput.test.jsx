import React from 'react'
import renderer from 'react-test-renderer'

import TextInput from './TextInput'

test('TextInput renders correctly', () => {
  const component = renderer.create(<TextInput htmlId="test" name="test" label="Test" onChange={() => {}} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('TextInput renders correctly when passed required prop', () => {
  const component = renderer.create(<TextInput htmlId="test" name="test" label="Test" required onChange={() => {}} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('TextInput renders correctly when passed an error prop', () => {
  const component = renderer.create(<TextInput htmlId="test" name="test" label="Test" required onChange={() => {}} error="This is a required field" />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('TextInput renders correctly when passed type of email', () => {
  const component = renderer.create(<TextInput htmlId="test" name="test" label="Test" type="email" onChange={() => {}} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('TextInput renders correctly when passed type of number', () => {
  const component = renderer.create(<TextInput htmlId="test" name="test" label="Test" type="number" onChange={() => {}} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('TextInput renders correctly when passed type of tel', () => {
  const component = renderer.create(<TextInput htmlId="test" name="test" label="Test" type="tel" onChange={() => {}} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('TextInput renders correctly when passed type of password', () => {
  const component = renderer.create(<TextInput htmlId="test" name="test" label="Test" type="password" onChange={() => {}} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('TextInput renders correctly when passed size of xs', () => {
  const component = renderer.create(<TextInput htmlId="test" name="test" label="Test" size="xs" onChange={() => {}} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('TextInput renders correctly when passed size of xs', () => {
  const component = renderer.create(<TextInput htmlId="test" name="test" label="Test" size="xs" onChange={() => {}} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('TextInput renders correctly when passed size of sm', () => {
  const component = renderer.create(<TextInput htmlId="test" name="test" label="Test" size="sm" onChange={() => {}} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('TextInput renders correctly when passed size of md', () => {
  const component = renderer.create(<TextInput htmlId="test" name="test" label="Test" size="md" onChange={() => {}} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('TextInput renders correctly when passed size of lg', () => {
  const component = renderer.create(<TextInput htmlId="test" name="test" label="Test" size="lg" onChange={() => {}} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('TextInput renders correctly when passed size of xl', () => {
  const component = renderer.create(<TextInput htmlId="test" name="test" label="Test" size="xl" onChange={() => {}} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
