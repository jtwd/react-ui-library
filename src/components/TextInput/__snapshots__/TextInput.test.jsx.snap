// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput renders correctly 1`] = `
<div
  className="sc-ifAKCX ffCDom"
>
  <label
    className="sc-bdVaJa etThJd"
    htmlFor="test"
  >
    Test
     
  </label>
  <input
    className="sc-EHOje cUJljk"
    id="test"
    name="test"
    onChange={[Function]}
    placeholder=""
    type="text"
    value={undefined}
  />
</div>
`;

exports[`TextInput renders correctly when passed an error prop 1`] = `
<div
  className="sc-ifAKCX ffCDom"
>
  <label
    className="sc-bdVaJa etThJd"
    htmlFor="test"
  >
    Test
     
    <span
      className="sc-bwzfXH dYARQs"
      title="Required field"
    />
  </label>
  <input
    className="sc-EHOje cQrQFB"
    id="test"
    name="test"
    onChange={[Function]}
    placeholder=""
    type="text"
    value={undefined}
  />
  <div
    className="sc-bxivhb krvqla"
  >
    <i
      className="sc-htpNat giXiHP"
    >
      <svg
        fill="currentColor"
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "color": undefined,
            "verticalAlign": "middle",
          }
        }
        viewBox="0 0 40 40"
        width="1em"
      >
        <g>
          <path
            d="m24.4 27.9v5q0 0.5-0.4 1t-1 0.4h-5.7q-0.6 0-1-0.4t-0.4-1v-5q0-0.6 0.4-1t1-0.5h5.7q0.6 0 1 0.5t0.4 1z m0.7-23.6l-0.6 17.1q0 0.6-0.5 1t-1 0.5h-5.7q-0.6 0-1-0.5t-0.5-1l-0.6-17.1q0-0.6 0.4-1t1-0.4h7.1q0.6 0 1 0.4t0.4 1z"
          />
        </g>
      </svg>
    </i>
    This is a required field
  </div>
</div>
`;

exports[`TextInput renders correctly when passed required prop 1`] = `
<div
  className="sc-ifAKCX ffCDom"
>
  <label
    className="sc-bdVaJa etThJd"
    htmlFor="test"
  >
    Test
     
    <span
      className="sc-bwzfXH dYARQs"
      title="Required field"
    />
  </label>
  <input
    className="sc-EHOje cUJljk"
    id="test"
    name="test"
    onChange={[Function]}
    placeholder=""
    type="text"
    value={undefined}
  />
</div>
`;

exports[`TextInput renders correctly when passed type of email 1`] = `
<div
  className="sc-ifAKCX ffCDom"
>
  <label
    className="sc-bdVaJa etThJd"
    htmlFor="test"
  >
    Test
     
  </label>
  <input
    className="sc-EHOje cUJljk"
    id="test"
    name="test"
    onChange={[Function]}
    placeholder=""
    type="email"
    value={undefined}
  />
</div>
`;

exports[`TextInput renders correctly when passed type of number 1`] = `
<div
  className="sc-ifAKCX ffCDom"
>
  <label
    className="sc-bdVaJa etThJd"
    htmlFor="test"
  >
    Test
     
  </label>
  <input
    className="sc-EHOje cUJljk"
    id="test"
    name="test"
    onChange={[Function]}
    placeholder=""
    type="number"
    value={undefined}
  />
</div>
`;

exports[`TextInput renders correctly when passed type of password 1`] = `
<div
  className="sc-ifAKCX ffCDom"
>
  <label
    className="sc-bdVaJa etThJd"
    htmlFor="test"
  >
    Test
     
  </label>
  <input
    className="sc-EHOje cUJljk"
    id="test"
    name="test"
    onChange={[Function]}
    placeholder=""
    type="password"
    value={undefined}
  />
</div>
`;

exports[`TextInput renders correctly when passed type of tel 1`] = `
<div
  className="sc-ifAKCX ffCDom"
>
  <label
    className="sc-bdVaJa etThJd"
    htmlFor="test"
  >
    Test
     
  </label>
  <input
    className="sc-EHOje cUJljk"
    id="test"
    name="test"
    onChange={[Function]}
    placeholder=""
    type="tel"
    value={undefined}
  />
</div>
`;
